seqs/seqs.gobra

```diff
@@ -0,0 +1,396 @@
+	return let _ := DropFirstPrependFirst(xs) in
+	(IsSingleton(xs) ? First(xs) : math.Max(First(xs), Max(DropFirst(xs))))
+}
+
+// The maximum of the concatenation of two non-empty sequences is greater than
+// or equal to the maxima of its two non-empty subsequences.
+ghost
+requires len(xs) > 0 && len(ys) > 0
+ensures Max(xs ++ ys) >= Max(xs)
+ensures Max(xs ++ ys) >= Max(ys)
+ensures forall k int :: { k in xs ++ ys } (k in (xs ++ ys)) ==> Max(xs ++ ys) >= k
+decreases xs
+pure func MaxOfConcat(xs, ys seq[int]) util.Unit {
+	return let _ := ConcatDropFirst(xs, ys) in
+	(IsSingleton(xs) ? util.Unit{} : MaxOfConcat(DropFirst(xs), ys))
+}
+
+// Returns the minimum of a non-empty sequence.
+ghost
+requires len(xs) > 0
+ensures forall k int :: { k in xs } (k in xs) ==> result <= k
+ensures result in xs
+decreases xs
+pure func Min(xs seq[int]) (result int) {
+	return let _ := DropFirstPrependFirst(xs) in
+	(IsSingleton(xs) ? First(xs) : math.Min(First(xs), Min(DropFirst(xs))))
+}
+
+// The minimum of the concatenation of two non-empty sequences is less than or
+// equal to the minima of its two non-empty subsequences.
+ghost
+requires len(xs) > 0 && len(ys) > 0
+ensures Min(xs ++ ys) <= Min(xs)
+ensures Min(xs ++ ys) <= Min(ys)
+ensures forall k int :: { k in xs ++ ys } (k in xs ++ ys) ==> Min(xs ++ ys) <= k
+decreases xs
+pure func MinOfConcat(xs, ys seq[int]) util.Unit {
+	return let _ := ConcatDropFirst(xs, ys) in
+	(IsSingleton(xs) ? util.Unit{} : MinOfConcat(DropFirst(xs), ys))
+}
+
+// The maximum element in a non-empty sequence is greater than or equal to the
+// maxima of its non-empty subsequences.
+ghost
+requires 0 <= a && a < b && b <= len(xs)
+ensures Max(xs[a:b]) <= Max(xs)
+decreases
+pure func MaxOfSubseq(xs seq[int], a, b int) util.Unit {
+	return util.Asserting(Max(xs[a:b]) in xs)
+}
+
+// The minimum element in a non-empty sequence is less than or equal to the
+// minima of its non-empty subsequences.
+ghost
+requires 0 <= a && a < b && b <= len(xs)
+ensures Min(xs[a:b]) >= Min(xs)
+decreases
+pure func MinOfSubseq(xs seq[int], a, b int) util.Unit {
+	return util.Asserting(Min(xs[a:b]) in xs)
+}
+
+// Flattens a sequence of sequences into a single sequence.
+ghost
+decreases xs
+pure func Flatten(xs seq[seq[int]]) seq[int] {
+	return len(xs) == 0 ? Empty() : xs[0] ++ Flatten(xs[1:])
+}
+
+// Flatten is distributive over concatenation.
+ghost
+ensures Flatten(xs ++ ys) == Flatten(xs) ++ Flatten(ys)
+decreases xs
+pure func FlattenConcat(xs, ys seq[seq[int]]) util.Unit {
+	return len(xs) == 0 ? util.Unit{} :
+	let _ := util.Asserting((xs ++ ys)[1:] == xs[1:] ++ ys) in
+	FlattenConcat(xs[1:], ys)
+}
+
+// Flattening a sequence containing a single sequence yields the sequence within.
+ghost
+ensures Flatten(seq[seq[int]]{xs}) == xs
+decreases
+pure func FlattenSingleton(xs seq[int]) util.Unit {
+	// QUES Why do I need to assert the definition of Flatten? assert len(seq[seq[int]]{xs}) != 0 isn't enough.
```

Comment:
I also wonder if we need to assert the definition of `Flatten`. Could you try this with the following body?
```go
pure func FlattenSingleton(xs seq[int]) util.Unit {
    return let _ := Flatten(seq[seq[int]]{xs}) in Unit{}
}
```
It might be that the quantifier in the definitional axiom is not being triggered, and adding the call to `Flatten` might solve this. I would be curious anyway.

Reply:
Here is the minimized Viper program:

```
domain Unit {
  function unit(): Unit
}

function Asserting(b: Bool): Unit
  requires b
{ unit() }

function Flatten(xs: Seq[Seq[Int]]): Seq[Int]
  decreases xs
{
  (|xs| == 0 ?
    Seq[Int]() :
    xs[0] ++ Flatten(xs[1..]))
}

function FlattenSingleton(xs: Seq[Int]): Unit
  ensures Flatten(Seq(xs)) == xs
{
  // Does not work:
  // unit()
  
  // Does not work:
  // Asserting(|Seq(xs)| != 0)

  // Does not work:
  // let _ == (Flatten(Seq(xs))) in unit()

  // This works:
  Asserting(Flatten(Seq(xs)) == (|Seq(xs)| == 0 ?
    Seq[Int]() :
    Seq(xs)[0] ++ Flatten(Seq(xs)[1..])))
}
```
Do you think this is a Silicon issue?

Reply:
I think it may be. I would open an issue on the silicon issue tracker